### PR TITLE
JAMES-3312: Integration test for GET endpoint to retrieve the Session object

### PR DESCRIPTION
**Pull request by nvduc91**
Integration tests for GET endpoint to retrieve the Session object

### CODE REVIEW

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/SessionRoutesContract.scala [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        .contentType(JSON)
+      .extract()
+        .body()
+        .asString()
+
+    assertThatJson(Json.parse(sessionJson)).isEqualTo(session_object_json)
+  }
+
+  @Test
+  def getWithoutRFC8621HeaderShouldReturn404() {
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/SessionRoutesContract.scala:129]: This test is imo not needed.

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/SessionRoutesContract.scala [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+      .addUser(BOB.asString, BOB_PASSWORD)
+      .addUser(ANDRE.asString, ANDRE_PASSWORD)
+
+    requestSpecification = baseRequestSpecBuilder(server)
+        .setAuth(authScheme(UserCredential(BOB, BOB_PASSWORD)))
+      .build
+  }
+
+  @Test
+  def getShouldReturnCorrectSession() {
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/SessionRoutesContract.scala:113]: This is a basic feature

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/SessionRoutesContract.scala [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+  def getWithoutRFC8621HeaderShouldReturn404() {
+    `given`()
+      .when()
+        .get("/session")
+      .`then`
+        .statusCode(SC_NOT_FOUND)
+  }
+
+  @Test
+  def wrongHttpMethodShouldReturn404() {
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/SessionRoutesContract.scala:138]: Not needed too imo

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/SessionRoutesContract.scala [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+                         |    "urn:ietf:params:jmap:mail" : "0fe275bf13ff761407c17f64b1dfae2f4b3186feea223d7267b79f873a105401"
+                         |  },
+                         |  "username" : "bob@james.org",
+                         |  "apiUrl" : "http://this-url-is-hardcoded.org/jmap",
+                         |  "downloadUrl" : "http://this-url-is-hardcoded.org/download",
+                         |  "uploadUrl" : "http://this-url-is-hardcoded.org/upload",
+                         |  "eventSourceUrl" : "http://this-url-is-hardcoded.org/eventSource",
+                         |  "state" : "000001"
+                         |}""".stripMargin
+  private val session_object_json = Json.parse(expected_session_object);
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/SessionRoutesContract.scala:95]: Reuse the ones you declared in `Fixture`

File: server/protocols/jmap-rfc-8621/src/test/scala/org/apache/james/jmap/json/Fixture.scala [hunk 5]
```diff
@@ -23,7 +23,7 @@ import eu.timepit.refined.auto._
 import org.apache.james.jmap.model.CapabilityIdentifier.CapabilityIdentifier
 import org.apache.james.jmap.model.Id.Id
 import org.apache.james.jmap.model.Invocation.{Arguments, MethodCallId, MethodName}
-import org.apache.james.jmap.model.{ClientId, CreatedIds, Invocation, ResponseObject, ServerId}
+import org.apache.james.jmap.model._
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621/src/test/scala/org/apache/james/jmap/json/Fixture.scala:26]: Keep the detailed list of imports

File: server/protocols/jmap-rfc-8621/src/test/scala/org/apache/james/jmap/json/Fixture.scala [hunk 6]
```diff
@@ -47,4 +47,65 @@ object Fixture {
   val responseObject1: ResponseObject = ResponseObject(ResponseObject.SESSION_STATE, Seq(invocation1))
   val responseObject2: ResponseObject = ResponseObject(ResponseObject.SESSION_STATE, Seq(invocation2))
   val responseObjectWithUnsupportedMethod: ResponseObject = ResponseObject(ResponseObject.SESSION_STATE, Seq(invocation1, unsupportedInvocation))
+
+  val session_object_json = Json.parse(expected_session_object);
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621/src/test/scala/org/apache/james/jmap/json/Fixture.scala:51]: Would make more sense to have this declared below `expected_session_object`

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/SessionRoutesContract.scala (hunk 4 above)
Comment (nvduc91) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/SessionRoutesContract.scala:95]: It needed to import a test-jar for it and it not worth IMO

**Comment by chibenwa** [comment]
```
[4e2e013bf3615e9349238425f095740ed7e23688] [INFO] -------------------------------------------------------
[4e2e013bf3615e9349238425f095740ed7e23688] [INFO]  T E S T S
[4e2e013bf3615e9349238425f095740ed7e23688] [INFO] -------------------------------------------------------
[4e2e013bf3615e9349238425f095740ed7e23688] [INFO] Running org.apache.james.jmap.http.UserCredentialParserTest
[4e2e013bf3615e9349238425f095740ed7e23688] [INFO] Tests run: 13, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.472 s - in org.apache.james.jmap.http.UserCredentialParserTest
[4e2e013bf3615e9349238425f095740ed7e23688] [INFO] 
[4e2e013bf3615e9349238425f095740ed7e23688] [INFO] Results:
[4e2e013bf3615e9349238425f095740ed7e23688] [INFO] 
[4e2e013bf3615e9349238425f095740ed7e23688] [INFO] Tests run: 13, Failures: 0, Errors: 0, Skipped: 0
[4e2e013bf3615e9349238425f095740ed7e23688] [INFO] 
[4e2e013bf3615e9349238425f095740ed7e23688] [INFO] 
[4e2e013bf3615e9349238425f095740ed7e23688] [INFO] --- scalatest-maven-plugin:2.0.0:test (test) @ james-server-jmap-rfc-8621 ---
[4e2e013bf3615e9349238425f095740ed7e23688] Discovery starting.
[4e2e013bf3615e9349238425f095740ed7e23688] Discovery completed in 1 second, 200 milliseconds.
[4e2e013bf3615e9349238425f095740ed7e23688] Run starting. Expected test count is: 76
[4e2e013bf3615e9349238425f095740ed7e23688] SessionSupplierTest:
[4e2e013bf3615e9349238425f095740ed7e23688] generate
[4e2e013bf3615e9349238425f095740ed7e23688] - should return correct username
[4e2e013bf3615e9349238425f095740ed7e23688]   should return correct account which
[4e2e013bf3615e9349238425f095740ed7e23688]   - has size
[4e2e013bf3615e9349238425f095740ed7e23688]   - has name
[4e2e013bf3615e9349238425f095740ed7e23688]   - has accountId being hash of username in string
[4e2e013bf3615e9349238425f095740ed7e23688] MailboxValidationTest:
[4e2e013bf3615e9349238425f095740ed7e23688] MailboxValidation
[4e2e013bf3615e9349238425f095740ed7e23688] - should succeed
[4e2e013bf3615e9349238425f095740ed7e23688] - should fail when mailboxPath is invalid
[4e2e013bf3615e9349238425f095740ed7e23688] - should fail when unreadEmails is invalid
[4e2e013bf3615e9349238425f095740ed7e23688] - should fail when unreadThreads is invalid
[4e2e013bf3615e9349238425f095740ed7e23688] - should fail when totalEmails is invalid
[4e2e013bf3615e9349238425f095740ed7e23688] - should fail when totalThreads is invalid
[4e2e013bf3615e9349238425f095740ed7e23688] RequestObjectSerializationTest:
[4e2e013bf3615e9349238425f095740ed7e23688] Deserialize RequestObject
[4e2e013bf3615e9349238425f095740ed7e23688] *** RUN ABORTED ***
[4e2e013bf3615e9349238425f095740ed7e23688]   java.lang.ExceptionInInitializerError:
[4e2e013bf3615e9349238425f095740ed7e23688]   at org.apache.james.jmap.json.RequestObjectSerializationTest.$anonfun$new$2(RequestObjectSerializationTest.scala:38)
[4e2e013bf3615e9349238425f095740ed7e23688]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[4e2e013bf3615e9349238425f095740ed7e23688]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[4e2e013bf3615e9349238425f095740ed7e23688]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[4e2e013bf3615e9349238425f095740ed7e23688]   at org.scalatest.Transformer.apply(Transformer.scala:22)
[4e2e013bf3615e9349238425f095740ed7e23688]   at org.scalatest.Transformer.apply(Transformer.scala:20)
[4e2e013bf3615e9349238425f095740ed7e23688]   at org.scalatest.wordspec.AnyWordSpecLike$$anon$3.apply(AnyWordSpecLike.scala:1076)
[4e2e013bf3615e9349238425f095740ed7e23688]   at org.scalatest.TestSuite.withFixture(TestSuite.scala:196)
[4e2e013bf3615e9349238425f095740ed7e23688]   at org.scalatest.TestSuite.withFixture$(TestSuite.scala:195)
[4e2e013bf3615e9349238425f095740ed7e23688]   at org.scalatest.wordspec.AnyWordSpec.withFixture(AnyWordSpec.scala:1879)
[4e2e013bf3615e9349238425f095740ed7e23688]   ...
[4e2e013bf3615e9349238425f095740ed7e23688]   Cause: java.lang.NullPointerException:
[4e2e013bf3615e9349238425f095740ed7e23688]   at com.fasterxml.jackson.core.JsonFactory.createParser(JsonFactory.java:889)
[4e2e013bf3615e9349238425f095740ed7e23688]   at play.api.libs.json.jackson.JacksonJson$.parseJsValue(JacksonJson.scala:285)
[4e2e013bf3615e9349238425f095740ed7e23688]   at play.api.libs.json.StaticBinding$.parseJsValue(StaticBinding.scala:17)
[4e2e013bf3615e9349238425f095740ed7e23688]   at play.api.libs.json.Json$.parse(Json.scala:173)
[4e2e013bf3615e9349238425f095740ed7e23688]   at org.apache.james.jmap.json.Fixture$.<clinit>(Fixture.scala:51)
[4e2e013bf3615e9349238425f095740ed7e23688]   at org.apache.james.jmap.json.RequestObjectSerializationTest.$anonfun$new$2(RequestObjectSerializationTest.scala:38)
[4e2e013bf3615e9349238425f095740ed7e23688]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[4e2e013bf3615e9349238425f095740ed7e23688]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[4e2e013bf3615e9349238425f095740ed7e23688]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[4e2e013bf3615e9349238425f095740ed7e23688]   at org.scalatest.Transformer.apply(Transformer.scala:22)
[4e2e013bf3615e9349238425f095740ed7e23688]   ...
```

Sounds related. Please @nvduc91 have a look...

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/SessionRoutesContract.scala (hunk 4 above)
Comment (Arsnael) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/SessionRoutesContract.scala]: Ah yeah my bad, different packages

File: server/protocols/jmap-rfc-8621/src/test/scala/org/apache/james/jmap/json/Fixture.scala (hunk 6 above)
Comment (Arsnael) [reply server/protocols/jmap-rfc-8621/src/test/scala/org/apache/james/jmap/json/Fixture.scala]: Not solved

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/SessionRoutesContract.scala [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+                         |      "collationAlgorithms" : [ "i;unicode-casemap" ]
+                         |    },
+                         |    "urn:ietf:params:jmap:mail" : {
+                         |      "maxMailboxesPerEmail" : 10000000,
+                         |      "maxMailboxDepth" : null,
+                         |      "maxSizeMailboxName" : 200,
+                         |      "maxSizeAttachmentsPerEmail" : 20000000,
+                         |      "emailQuerySortOptions" : [ "receivedAt", "cc", "from", "to", "subject", "size", "sentAt", "hasKeyword", "uid", "Id" ],
+                         |      "mayCreateTopLevelMailbox" : true
+                         |    }
```
Comment (rouazana) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/SessionRoutesContract.scala:55]: where are other supported capabilities?

File: server/protocols/jmap-rfc-8621/src/test/scala/org/apache/james/jmap/json/Fixture.scala [hunk 8]
```diff
@@ -47,4 +47,65 @@ object Fixture {
...
+                         |      "collationAlgorithms" : [ "i;unicode-casemap" ]
+                         |    },
+                         |    "urn:ietf:params:jmap:mail" : {
+                         |      "maxMailboxesPerEmail" : 10000000,
+                         |      "maxMailboxDepth" : null,
+                         |      "maxSizeMailboxName" : 200,
+                         |      "maxSizeAttachmentsPerEmail" : 20000000,
+                         |      "emailQuerySortOptions" : [ "receivedAt", "cc", "from", "to", "subject", "size", "sentAt", "hasKeyword", "uid", "Id" ],
+                         |      "mayCreateTopLevelMailbox" : true
+                         |    }
```
Comment (rouazana) [inline server/protocols/jmap-rfc-8621/src/test/scala/org/apache/james/jmap/json/Fixture.scala:70]: (idem)

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/SessionRoutesContract.scala (hunk 7 above)
Comment (nvduc91) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/SessionRoutesContract.scala:55]: It should not related to this PR, i will make another PR then rebase on this.
Please give me those information to handle the missing ones.

Comment (rouazana) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/SessionRoutesContract.scala:55]: ok see https://github.com/linagora/james-project/issues/3609

**Comment by chibenwa** [comment]
`java.lang.RuntimeException: java.io.IOException: Failed to bind to /0.0.0.0:43683`

test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**